README.md
# OpenPipelineIO
![example workflow](https://github.com/digital-idea/OpenPipelineIO/actions/workflows/go.yml/badge.svg)
[![Go Report Card](https://goreportcard.com/badge/github.com/digital-idea/OpenPipelineIO)](https://goreportcard.com/report/github.com/digital-idea/OpenPipelineIO)

![screenshot](figures/screenshot.png)

![review](figures/review.png)

![statistics](figures/statistics.png)

OpenPipelineIO는 영상(영화,드라마), 게임, 에니메이션등 콘텐츠 제작을 위한 프로젝트 매니징 솔루션, DATA IO 솔루션 입니다.

- 속도, 검색어 방식, 교육의 최소화, 단일파일 배포를 중점으로 개발되고 있습니다.
- 내부, 외부 서버에 설치가 가능합니다.
- 리뷰 시스템
- 사용자별 토큰키, 암호화키, 직급별 접근권한 사용이 가능합니다.

## Roadmap

브라우저 지원 우선순위: 크롬 > 사파리 > 파이어폭스

1. 파트너 관리툴 추가 - 진행중
1. 통계, 레포트 - 진행중
1. 달력, 간트챠트 - 리서치중
1. InBox 구조: 메시지, 팔로우 기능
1. 장비관리 툴
1. 웹 스캔 툴: 도토리의 multi clip, multi footage 연동
1. 소프트웨어 등록, 환경변수 관리 -> JWT토큰 정보를 uri로 보내는 형태의 웹프로토콜
1. 샷 파일 브라우져

### 데이터베이스(mongoDB) 설치 및 서비스 실행

CentOS

- [CentOS7에서 mongoDB 설정](https://github.com/cgiseminar/curriculum/blob/master/docs/install_mongodb.md)

macOS

```bash
brew uninstall mongodb
brew tap mongodb/brew
brew install mongodb-community
brew services start mongodb-community
```

Windows

- <https://fastdl.mongodb.org/win32/mongodb-win32-x86_64-2008plus-ssl-4.0.10-signed.msi>
- Download: <https://www.mongodb.com/download-center/community?jmp=docs>
- Setup: <https://docs.mongodb.com/manual/tutorial/install-mongodb-on-windows-unattended/>

### 실행하기

```bash
sudo openpipelineio -http :80 # 웹서버를 실행합니다.
sudo openpipelineio -http :80 -reviewrender # 웹서버 및 FFmpeg를 이용하여 리뷰를 렌더링하는 서버
sudo openpipelineio -http :80 &> openpipelineio.log& # openpipelineio.log 파일에 로그가 생성되면서 서비스 실행
```

> 여러분이 macOS를 사용한다면 기본적으로 80포트는 아파치 서버가 사용중일 수 있습니다. 80포트에 실행되는 아파치 서버를 종료하기 위해서 `$ sudo apachectl stop` 를 터미널에 입력해주세요.

OpenPipelineIO는 [wfs-웹파일시스템](https://github.com/digital-idea/wfs), [dilog-로그서버](https://github.com/digital-idea/dilog), [dilink-웹프로토콜](https://github.com/digital-idea/dilink)과 같이 연동됩니다. 아래 서비스 실행 및 프로토콜 설치도 같이 진행하면 더욱 편리한 OpenPipelineIO를 활용할 수 있습니다.

```bash
dilog -http :8080
wfs -http :8081
```

### CentOS 방화벽 설정

다른 컴퓨터에서 접근하기 위해서는 해당 포트를 방화벽 해제합니다.

```bash
sudo firewall-cmd --zone=public --add-port=80/tcp --permanent
success
sudo firewall-cmd --reload
```

### CommandLine

터미널에서 간단하게 명령어를 통해 관리를 할 수 있습니다.

- [Item](documents/item.md): Asset, Shot
- [User](documents/user.md)
- [Daily](documents/daily.md)
- [Organization](documents/organization.md)

### RestAPI

OpenPipelineIO는 RestAPI가 설계되어 있습니다.
Python, Go, Java, Javascript, node.JS, C++, C, C# 등 수많은 언어를 활용하여 OpenPipelineIO를 제어할 수 있습니다.

- [Project](documents/rest_project.md)
- [Item](documents/rest_item.md): Asset, Shot
- [User](documents/rest_user.md)
- [Organization](documents/rest_organization.md)
- [Tasksetting](documents/rest_tasksetting.md)
- [Status](documents/rest_status.md)
- [Review](documents/rest_review.md)
- [Statistics](documents/rest_statistics.md)
- [Partner](documents/rest_partner.md)
- [ProjectForPartner](documents/rest_projectforpartner.md)
- [Money](documents/rest_money.md)
- [Moneytype](documents/rest_moneytype.md)
- [Step](documents/rest_step.md)
- [Pipelinestep](documents/rest_pipelinestep.md)

### 썸네일 경로

위에서 생성된 thumbnail 폴더는 아래 구조를 띄고 있습니다.
썸네일은 사내 다른 응용프로그램에서도 사용될 수 있기 때문에 경로구조를 표기해둡니다.

- 썸네일주소 : `thumbnail/{projectname}/{id}.jpg`
- 사용자이미지 : `thumbnail/user/{id}.jpg`

### 프로젝트 Process

- [디자인 프로세스](documents/process_designer.md)
- [개발 프로세스](documents/process_developer.md)
- [Onset Setellite](documents/setellite.md)
- [DB관리](documents/dbbackup.md)

### Developer

- OpenPipelineIO: <https://openpipeline.io>
- Log서버: <https://openpipeline.io:8080>
- WFS서버: <https://openpipeline.io:8081>
- 회사 전용 빌드문의: hello@lazypic.org
- Maintainer: Jason / jason@lazypic.org
- Committer: Alex / alex@lazypic.org
- Contributors:
- 체험계정 ID/PW: guest
  - Guest 계정은 모든 메뉴가 보이지 않습니다.
  - Guest 계정은 일부 기능만 테스트 가능한 모드입니다.
  - 만약 많은 기능을 테스트하고 싶다면 가입한 ID와 함께 권한변경 요청메일을 hello@lazypic.org로 보내주세요.

### Infomation

- [OpenPipelineIO의 역사](documents/history.md)
- License: BSD 3-Clause License

### License

- OpenPipelineIO: BSD 3-Clause License
- [JScolor](http://jscolor.com/download/): GNU GPL license v3
- [Dropzone](https://www.dropzonejs.com): MIT License
- [JQuery](https://jquery.org/license/): MIT license
- [VFS](https://github.com/blang/vfs): MIT license
- [HttpFS](https://github.com/shurcooL/httpfs): MIT license
- [VFSgen](https://github.com/shurcooL/vfsgen): MIT license
- [Excelize](https://github.com/360EntSecGroup-Skylar/excelize): BSD 3-Clause License
- [Slack go webhook](https://github.com/ashwanthkumar/slack-go-webhook): Apache License, Version 2.0
- [Captcha](https://github.com/dchest/captcha): Apache License, Version 2.0
- [Mgo](https://github.com/go-mgo/mgo): <https://github.com/go-mgo/mgo/blob/v2-unstable/LICENSE>
- [JWT go](https://github.com/golang-jwt/jwt): MIT license
- [OpenColorIO](https://github.com/AcademySoftwareFoundation/OpenColorIO): BSD 3-Clause License
- [alfg/mp4](https://github.com/alfg/mp4): MIT license
- [amarburg/go-quicktime](https://github.com/amarburg/go-quicktime): MIT license
- [Gollia Mux](https://github.com/gorilla/mux): BSD 3-Clause License


documents/history.md
# History

- '22. 8. 20: CSI -> OpenPipelineIO 로 이름변경
- '22. 2. 19: 기존 (주)투썬디지털아이디어 -> (주)바른손디지털아이디어 로 변경
- '20. 11: 사내 Review Process Test, 기능 업데이트
- '20. 7: [(주)투썬디지털아이디어](http://www.digitalidea.co.kr) + [Lazypic,LLC.(유한책임회사 레이지픽)](https://lazypic.org) 공동개발 시작
- '20. 5: 리뷰기능 개발시작
- '20. 4: 퍼블리쉬 기능지원
- '20. 1 ~ 3: 기존 스탠드얼론 툴 제거. 다크모드 지원. UX변경
- '19. 8. 30: 리서치: 웹용 PM 입력기(MultiTask로 변경되면 기존 X,Y축 입력방식이 아닌 웹입력기 제작이 필요. 기술테스트 완료)
- '19. 6 ~ 8: 로그인, 권한레벨(보안), 사용자 자료구조 추가
- '19. 7: 기존 CSI2 RestAPI 48개 함수 추가
- '19. 6 ~ 7: 빌드환경구축. 바이너리 파일하나로 배포하는 모델 구축 VFS 기술적용
- '19. 6. 5: Open-source 전환, 프리랜서 그룹 Lazypic과 기술협업 시작
- '18. 4 ~ '19. 5: 개발중단
- '18. 3. 25: 현장툴 Setellite 연결
- '17.10.13: 특허출원 거절. DOI: <https://doi.org/10.8080/1020160035391>

    ![qr](figures/qrcode.png)

- ’17. 5.29 ~ : 웹을 이용한 restAPI 개발시작.
- '17. 2.28 ~ : 프로젝트 자료구조 추가. 각 그래픽스 툴(maya, nuke, katana...)와 연동
- ’16. 8. 22 ~ : 통계 페이지 작성 시작
- ’16. 8. 17 ~ : 회사 인원증가로 CSI 장애발생. CSI2 소스코드를 스테틱 언어 Go로 실제 변경 시작. CSI3 부터는 Go를 이용해서 제작. 기존 클라우드 기반에서 테스트한 코드 적용
- ’16. 5. 1 ~ : 애니메이션 사업부(디지아트), VFX사업부, 뉴미디어, BTL(홍보영상) 사업부 관련 개발 요구사항 수집. 사업축소로 인해 클라우드 테스트 프로젝트 진행
- '16. 3. 24 : 특허출원 준비. 담당: 특허법인 이룸리온. 전자부 부장 기술거래사, 기술가치 평가사 정승호
- ’15.12.22 ~ : CSI와 [WFS(Web file system)](https://github.com/digital-idea/wfs), 로그서버 [dilog](https://github.com/digital-idea/dilog) 제작 연동
- ’15.10.28 ~ 17.1.22 : 웹서버 다운으로 스테틱언어로 변경. Go 테스트, 클라우드, 로그인테스트를 위한 테스트 프로젝트를 AWS로 진행.
- ’14.11.28 ~ ’16.1.28 : 영화 "몽키킹" 적용. 프로젝트를 진행하면서 불편했던 점 개선.
- ’13. 9. 4 ~ ’13.10.21 : CSI 3차 리뷰, 분석 / 구 파이프라인 개발자와 현재 자료구조 분석, 구 파이프라인툴(스파르타) 분석
- ’13. 7.31 ~ ’13. 8.25 : CSI 2차 리뷰, 분석
- ’13. 3.25 ~ : 영화 "히어로" 파일럿 적용
- ’12.11.8 ~ ’12.11.30 : CSI 프로젝트 시작, CSI 설계에 필요한 1차 분석.


main.go
package main

import (
	"bytes"
	"errors"
	"flag"
	"fmt"
	"html/template"
	"log"
	"os"
	"os/exec"
	"os/user"
	"strings"
	"time"

	"github.com/digital-idea/dilog"
	"gopkg.in/mgo.v2"
)

var (
	// DBIP 값은 컴파일 단계에서 회사에 따라 값이 바뀐다.
	DBIP = "127.0.0.1"
	// DBPORT mongoDB 기본포트.
	DBPORT = ":27017"
	// DBNAME 값은 데이터베이스 이름이다.
	DBNAME = "csi" // legacy
	// WFS 값은 컴파일 단계에서 회사에 따라 값이 바뀐다.
	WFS = "http://127.0.0.1:8081"
	// DILOG 값은 컴파일 단계에서 회사에 따라 값이 바뀐다.
	DILOG = "http://127.0.0.1:8080"
	// DNS 값은 서비스 DNS 값입니다.
	DNS = "openpipeline.io"
	// MAILDNS 값은 컴파일 단계에서 회사에 따라 값이 바뀐다.
	MAILDNS = "lazypic.org"
	// COMPANY 값은 컴파일 단계에서 회사에 따라 값이 바뀐다.
	COMPANY = "Lazypic"
	// TEMPLATES 값은 웹서버 실행전 사용할 템플릿이다.
	TEMPLATES = template.New("")
	// SHA1VER  은 Git SHA1 값이다.
	SHA1VER = "26b300a004abae553650c924514dc550e7385c9e" // 첫번째 커밋
	// BUILDTIME 은 빌드타임 시간이다.
	BUILDTIME = "2012-11-08T10:00:00" // 최초로 만든 시간
	// CachedAdminSetting 은 서비스 시작전 어드민 셋팅값을 메모리에 넣어서 사용되는 변수이다.
	CachedAdminSetting = Setting{}

	// 주요서비스 인수
	flagDBIP       = flag.String("dbip", DBIP+DBPORT, "mongodb ip and port")                                                            // mgo용 mongoDB 주소
	flagMongoDBURI = flag.String("mongodburi", fmt.Sprintf("mongodb://%s%s", DBIP, DBPORT), "mongoDB URI ex)mongodb://localhost:27017") //mongo-driver용 인수
	flagDBName     = flag.String("dbname", DBNAME, "mongodb db name")                                                                   // mongoDB DB이름
	flagMailDNS    = flag.String("maildns", MAILDNS, "mail DNS name")

	flagDebug          = flag.Bool("debug", false, "디버그모드 활성화")
	flagDevmode        = flag.Bool("devmode", false, "dev mode")
	flagHTTPPort       = flag.String("http", "", "Web Service Port number.")          // 웹서버 포트
	flagCompany        = flag.String("company", COMPANY, "Web Service Port number.")  // 회사이름
	flagVersion        = flag.Bool("version", false, "Print Version")                 // 버전
	flagCookieAge      = flag.Int64("cookieage", 168, "cookie age (hour)")            // 기본 일주일(168시간)로 설정한다. 참고: MPAA 기준 4시간이다.
	flagThumbnailAge   = flag.Int("thumbnailage", 1, "thumbnail image age (seconds)") // 썸네일 업데이트 시간. 3600초 == 1시간
	flagAuthmode       = flag.Bool("authmode", false, "restAPI authorization active") // restAPI 이용시 authorization 활성화
	flagCertFullchanin = flag.String("certfullchanin", fmt.Sprintf("/etc/letsencrypt/live/%s/fullchain.pem", DNS), "certification fullchain path")
	flagCertPrivkey    = flag.String("certprivkey", fmt.Sprintf("/etc/letsencrypt/live/%s/privkey.pem", DNS), "certification privkey path")
	// Process
	flagProcessBufferSize = flag.Int("processbuffersize", 100, "process buffer size") // 최대 대기 리스트
	flagMaxProcessNum     = flag.Int("maxprocessnum", 4, "max process number")        // 최대 연산 갯수
	flagReviewRender      = flag.Bool("reviewrender", false, "ffmpeg를 이용해서 리뷰 렌더링을 허용하는 옵션")

	// RV
	flagRVPath = flag.String("rvpath", "/opt/rv-Linux-x86-64-7.0.0/bin/rv", "rvplayer path")
	flagPlay   = flag.Bool("play", false, "Play RV")
	// Etc Service
	flagDILOG = flag.String("dilog", DILOG, "dilog webserver url and port. ex) "+DILOG)
	flagWFS   = flag.String("wfs", WFS, "wfs webserver url and port. ex) "+WFS)

	// Commandline Args
	flagAdd                = flag.String("add", "", "add project, add item(shot, asset)")
	flagRm                 = flag.String("rm", "", "remove project, shot, asset, user")
	flagProject            = flag.String("project", "", "project name")
	flagName               = flag.String("name", "", "name")
	flagSeason             = flag.String("season", "", "season")
	flagEpisode            = flag.String("episode", "", "episode")
	flagNetflixID          = flag.String("netflixid", "", "netflix id")
	flagType               = flag.String("type", "", "type: org,left,asset,org1,src,src1,lsrc,rsrc")
	flagAssettags          = flag.String("assettags", "", "asset tags, 입력예) prop,char,env,prop,comp,plant,vehicle,component,group,assembly 형태로 입력")
	flagAssettype          = flag.String("assettype", "", "assettype: char,env,global,prop,comp,plant,vehicle,group") // 추후 삭제예정.
	flagHelp               = flag.Bool("help", false, "자세한 도움말을 봅니다.")
	flagDate               = flag.String("date", "", "Date. ex) 2016-12-06")
	flagThumbnailImagePath = flag.String("thumbnailimagepath", "", "Thumbnail image 경로")
	flagThumbnailMovPath   = flag.String("thumbnailmovpath", "", "Thumbnail mov 경로")
	flagPlatePath          = flag.String("platepath", "", "Plate 경로")
	// Commandline Args: User
	flagID                = flag.String("id", "", "user id")
	flagAccessLevel       = flag.Int("accesslevel", -1, "edit user Access Level")
	flagSignUpAccessLevel = flag.Int("signupaccesslevel", 3, "signup access level")
	// scan정보 추가. plate scan tool에서 데이터를 등록할 때 활용되는 옵션
	flagPlatesize       = flag.String("platesize", "", "스캔 플레이트 사이즈")
	flagTask            = flag.String("task", "", "태스크 이름. 예) model,mm.layout,ani,fx,mg,fur,sim,crowd,light,comp,matte,env")
	flagScantimecodein  = flag.String("scantimecodein", "00:00:00:00", "스캔 Timecode In")
	flagScantimecodeout = flag.String("scantimecodeout", "00:00:00:00", "스캔 Timecode Out")
	flagJusttimecodein  = flag.String("justtimecodein", "00:00:00:00", "Just구간 Timecode In")
	flagJusttimecodeout = flag.String("justtimecodeout", "00:00:00:00", "Just구간 Timecode Out")
	flagScanframe       = flag.Int("scanframe", 0, "스캔 총 프레임수")
	flagScanname        = flag.String("scanname", "", "스캔 폴더명")
	flagScanin          = flag.Int("scanin", -1, "스캔 In Frame")
	flagScanout         = flag.Int("scanout", -1, "스캔 Out Frame")
	flagJustin          = flag.Int("justin", -1, "Just In Frame")
	flagJustout         = flag.Int("justout", -1, "Just Out Frame")
	flagPlatein         = flag.Int("platein", -1, "플레이트 In Frame")
	flagPlateout        = flag.Int("plateout", -1, "플레이트 Out Frame")
	flagUpdateParent    = flag.Bool("updateparent", false, "org1,org2 형태의 재스캔 항목이라면 원본 org 정보를 업데이트 한다.")
)

func main() {
	log.SetFlags(log.LstdFlags | log.Lshortfile)
	log.SetPrefix("OpenPipelineIO: ")
	flag.Usage = usage
	flag.Parse()
	if *flagVersion {
		fmt.Println("buildTime:", BUILDTIME)
		fmt.Println("git SHA1:", SHA1VER)
		os.Exit(0)
	}

	user, err := user.Current()
	if err != nil {
		log.Fatal(err)
	}
	ip, err := serviceIP()
	if err != nil {
		log.Fatal(err)
	}
	if *flagAccessLevel != -1 && *flagID != "" {
		if user.Username != "root" {
			log.Fatal(errors.New("사용자의 레벨을 수정하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		u, err := getUser(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		err = rmToken(session, u.ID)
		if err != nil {
			log.Fatal(err)
		}
		u.AccessLevel = AccessLevel(*flagAccessLevel)
		err = setUser(session, u)
		if err != nil {
			log.Fatal(err)
		}
		err = addToken(session, u)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagRm == "division" && *flagID != "" { // division 삭제
		if user.Username != "root" {
			log.Fatal(errors.New("사용자를 삭제하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		err = rmDivision(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagRm == "department" && *flagID != "" { // department 삭제
		if user.Username != "root" {
			log.Fatal(errors.New("사용자를 삭제하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		err = rmDepartment(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagRm == "team" && *flagID != "" { // team 삭제
		if user.Username != "root" {
			log.Fatal(errors.New("사용자를 삭제하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		err = rmTeam(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagRm == "role" && *flagID != "" { // role 삭제
		if user.Username != "root" {
			log.Fatal(errors.New("사용자를 삭제하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		err = rmRole(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagRm == "position" && *flagID != "" { // position 삭제
		if user.Username != "root" {
			log.Fatal(errors.New("사용자를 삭제하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		err = rmPosition(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagRm == "user" && *flagID != "" { // 사용자 삭제
		if user.Username != "root" {
			log.Fatal(errors.New("사용자를 삭제하기 위해서는 root 권한이 필요합니다"))
		}
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		u, err := getUser(session, *flagID)
		if err != nil {
			log.Fatal(err)
		}
		err = rmToken(session, u.ID)
		if err != nil {
			log.Fatal(err)
		}
		err = rmUser(session, u.ID)
		if err != nil {
			log.Fatal(err)
		}
		return
	} else if *flagAdd == "item" && *flagName != "" && *flagProject != "" && *flagType != "" { //아이템 추가
		switch *flagType {
		case "org", "left": // 일반영상은 org가 샷 타입이다. 입체프로젝트는 left가 샷타입이다.
			addShotItemCmd(*flagProject, *flagName, *flagType, *flagPlatesize, *flagScanname, *flagScantimecodein, *flagScantimecodeout, *flagJusttimecodein, *flagJusttimecodeout, *flagScanframe, *flagScanin, *flagScanout, *flagPlatein, *flagPlateout, *flagJustin, *flagJustout)
			dilog.Add(*flagDBIP, ip, "샷 생성되었습니다.", *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			dilog.Add(*flagDBIP, ip, "스캔이름 : "+*flagScanname, *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			dilog.Add(*flagDBIP, ip, fmt.Sprintf("스캔타임코드 : %s(%d) / %s(%d) (총%df)", *flagScantimecodein, *flagScanin, *flagScantimecodeout, *flagScanout, *flagScanframe), *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			dilog.Add(*flagDBIP, ip, fmt.Sprintf("플레이트 구간 : %d - %d", *flagPlatein, *flagPlateout), *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			dilog.Add(*flagDBIP, ip, "플레이트 사이즈 : "+*flagPlatesize, *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			return
		case "asset": //에셋 추가
			addAssetItemCmd(*flagProject, *flagName, *flagType, *flagAssettype, *flagAssettags)
			dilog.Add(*flagDBIP, ip, "에셋이 생성되었습니다.", *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			dilog.Add(*flagDBIP, ip, fmt.Sprintf("에셋타입 : %s, 에셋태그 : %s", *flagAssettype, *flagAssettags), *flagProject, *flagName+"_"+*flagType, "csi3", user.Username, 180)
			return
		default: //소스, 재스캔 추가
			addOtherItemCmd(*flagProject, *flagName, *flagType, *flagPlatesize, *flagScanname, *flagScantimecodein, *flagScantimecodeout, *flagJusttimecodein, *flagJusttimecodeout, *flagScanframe, *flagScanin, *flagScanout, *flagPlatein, *flagPlateout, *flagJustin, *flagJustout)
			logString := fmt.Sprintf("Create Item: %s_%s, Scanname: %s, ScanTimecode: %s(%d) / %s(%d) (Total:%df), Plate Range: %d - %d, Platesize: %s",
				*flagName,
				*flagType,
				*flagScanname,
				*flagScantimecodein, *flagScanin, *flagScantimecodeout, *flagScanout, *flagScanframe,
				*flagPlatein, *flagPlateout,
				*flagPlatesize,
			)
			dilog.Add(*flagDBIP, ip, logString, *flagProject, *flagName, "csi3", user.Username, 180)
			if *flagUpdateParent {
				// updateParent 옵션이 활성화되어있고, org, left가 재스캔이라면.. 원본플레이트의 정보를 업데이트한다.
				if (*flagType != "org" && strings.Contains(*flagType, "org")) || (*flagType != "left" && strings.Contains(*flagType, "left")) {
					session, err := mgo.Dial(*flagDBIP)
					if err != nil {
						log.Fatal(err)
					}
					defer session.Close()
					typ := "org"
					if strings.Contains(*flagType, "left") {
						typ = "left"
					}
					item, err := getItem(session, *flagProject, *flagName+"_"+typ)
					if err != nil {
						log.Fatal(err)
					}
					item.Platesize = *flagPlatesize
					item.ScanTimecodeIn = *flagScantimecodein
					item.ScanTimecodeOut = *flagScantimecodeout
					item.JustTimecodeIn = *flagJusttimecodein
					item.JustTimecodeOut = *flagJusttimecodeout
					item.ScanIn = *flagScanin
					item.ScanOut = *flagScanout
					item.ScanFrame = *flagScanframe
					item.Scanname = *flagScanname
					item.PlateIn = *flagPlatein
					item.PlateOut = *flagPlateout
					item.JustIn = *flagJustin
					item.JustOut = *flagJustout
					item.UseType = *flagType
					// adminsetting 값을 가지고와서 Thumbnailmov 값을 설정한다.
					admin, err := GetAdminSetting(session)
					if err != nil {
						log.Fatal(err)
					}
					var thumbnailMovPath bytes.Buffer
					thumbnailMovPathTmpl, err := template.New("thumbnailMovPath").Parse(admin.ThumbnailMovPath)
					if err != nil {
						log.Fatal(err)
					}
					err = thumbnailMovPathTmpl.Execute(&thumbnailMovPath, item)
					if err != nil {
						log.Fatal(err)
					}
					item.Thummov = thumbnailMovPath.String()

					err = setItem(session, *flagProject, item)
					if err != nil {
						log.Fatal(err)
					}
					// log
					logString := fmt.Sprintf("Scanname: %s, ScanTimecode: %s(%d) / %s(%d) Total: %df\nPlate Range: %d - %d\nPlatesize: %s",
						*flagScanname,
						*flagScantimecodein,
						*flagScanin,
						*flagScantimecodeout,
						*flagScanout,
						*flagScanframe,
						*flagPlatein,
						*flagPlateout,
						*flagPlatesize,
					)
					dilog.Add(*flagDBIP, ip, logString, *flagProject, *flagName, "csi3", user.Username, 180)
				}
			}
			return
		}
	} else if *flagRm == "item" && *flagName != "" && *flagProject != "" && *flagType != "" { //아이템 삭제
		rmItemCmd(*flagProject, *flagName, *flagType)
		return
	} else if *flagHTTPPort != "" {
		// 만약 프로젝트가 하나도 없다면 "TEMP" 프로젝트를 생성한다. 프로젝트가 있어야 템플릿이 작동하기 때문이다.
		session, err := mgo.DialWithTimeout(*flagDBIP, 2*time.Second)
		if err != nil {
			log.Fatal("DB가 실행되고 있지 않습니다.")
		}
		admin, err := GetAdminSetting(session)
		if err != nil {
			log.Fatal(err)
		}
		// 어드민설정을 한번 저장한다. CachedAdminSetting값은 매번 DB를 호출하면 안되는 작업에서 사용된다.
		CachedAdminSetting = admin
		// 만약 Admin설정에 ThumbnailRootPath가 잡혀있다면 그 값을 이용한다.
		if admin.ThumbnailRootPath == "" {
			log.Println("admin 설정창의 thumbnail 경로지정이 필요합니다.")
		}

		plist, err := Projectlist(session)
		if err != nil {
			log.Fatal(err)
		}
		// 프로젝트가 존재하지 않는다면 test 프로젝트를 추가한다.
		if len(plist) == 0 {
			p := *NewProject("test")
			err = addProject(session, p)
			if err != nil {
				log.Fatal(err)
			}
		}
		session.Close()
		if *flagHTTPPort == ":80" {
			fmt.Printf("Service start: http://%s\n", ip)
		} else if *flagHTTPPort == ":443" {
			fmt.Printf("Service start: https://%s\n", ip)
		} else {
			fmt.Printf("Service start: http://%s%s\n", ip, *flagHTTPPort)
		}
		vfsTempates, err := LoadTemplates()
		if err != nil {
			log.Fatal(err)
		}
		TEMPLATES = vfsTempates
		if *flagReviewRender {
			go ProcessMain() // 연산(Review데이터 등등)이 필요한 것들이 있다면 연산을 시작한다.
		}
		webserver(*flagHTTPPort)
	} else if MatchNormalTime.MatchString(*flagDate) {
		// date 값이 데일리 형식이면 해당 날짜에 업로드된 mov를 RV를 통해 플레이한다.
		// 예: $ openpipelineio -date 2020-03-19 -play
		session, err := mgo.Dial(*flagDBIP)
		if err != nil {
			log.Fatal(err)
		}
		defer session.Close()
		dbProjectlist, err := Projectlist(session)
		if err != nil {
			log.Fatal(err)
		}
		// 해당프로젝트만 데일리를 위한 옵션
		// 만약 담당 프로젝트 감독님이 오면 해당 프로젝트 영상만 띄운다.
		reviewProjectlist := dbProjectlist
		if *flagProject != "" {
			err := HasProject(session, *flagProject)
			if err != nil {
				log.Fatalln(err)
			}
		}
		var playlist []string
		for _, project := range reviewProjectlist {
			op := SearchOption{
				Project:    project,
				Searchword: *flagDate,
				Assign:     true,
				Ready:      true,
				Wip:        true,
				Confirm:    true,
				Done:       true,
				Out:        true,
				Sortkey:    "name",
			}
			items, err := Search(session, op)
			if err != nil {
				log.Fatalln(err)
			}
			// 만약 태스크명을 입력받았다면, 태스크명이 유효한지 체크하는 부분.
			if *flagTask != "" {
				hastask := false
				tasks, err := TasksettingNames(session)
				if err != nil {
					log.Fatal(err)
				}
				for _, t := range tasks {
					if *flagTask == strings.ToLower(t) {
						hastask = true
					}
				}
				if !hastask {
					log.Fatalf("%s Task 이름은 사용할 수 없습니다.\n", *flagTask)
				}
			}
			// 검색 옵션을 이용해서 daily 리스트를 만든다.
			for _, item := range items {
				tasks, err := TasksettingNames(session)
				if err != nil {
					log.Println(err)
					continue
				}
				for _, t := range tasks {
					if _, found := item.Tasks[t]; !found {
						continue
					}
					if *flagDate == ToNormalTime(item.Tasks[t].Mdate) && isMov(item.Tasks[t].Mov) {
						playlist = append(playlist, item.Tasks[t].Mov)
					}
				}
			}
		}
		// -play 인수가 붙어있다면, RV를 이용해서 플레이한다.
		if *flagPlay {
			// adminsetting에 RV 경로가 설정되어 있다면 해당 설정경로를 가지고 온다.
			admin, err := GetAdminSetting(session)
			if err != nil {
				log.Fatal(err)
			}
			if admin.RVPath != "" {
				*flagRVPath = admin.RVPath
			}
			if _, err := os.Stat(*flagRVPath); os.IsNotExist(err) {
				fmt.Println("RV가 rvpath에 존재하지 않습니다.")
				os.Exit(1)
			}
			out, err := exec.Command(*flagRVPath, playlist...).CombinedOutput()
			if err != nil {
				fmt.Printf("%v: %s\n", err, string(out))
				os.Exit(1)
			}
		}
		// -play 인수가 없다면, mov경로만 출력한다.
		for _, mov := range playlist {
			fmt.Println(mov)
		}
		return
	}
	if *flagHelp {
		flag.Usage()
		return
	}
	flag.Usage()
}
